client: move Transaction initialization into constructor

andreimatei opened the pull request:
Fixes #13894.
Fixes #16908.

Before this patch, we were initializing a Transaction proto the first
time client.Txn.Send() was called. This patch moves the initialization
of the proto to the client.Txn ctor. This means that a txn ID is
assigned early and so that all code can rely on transactions having IDs,
which simplifies things.
The transaction timestamp is also assigned early, and so
TxnExecOptions.AssignOrigTimestamp can go away.

andreimatei committed:
client: move Transaction initialization into constructor

Fixes #13894.
Fixes #16908.
Touches #16827.

Before this patch, we were initializing a Transaction proto the first
time client.Txn.Send() was called. This patch moves the initialization
of the proto to the client.Txn ctor. This means that a txn ID is
assigned early and so that all code can rely on transactions having IDs,
which simplifies things.
The transaction timestamp is also assigned early, and so
TxnExecOptions.AssignOrigTimestamp can go away.

andreimatei committed:
storage: remove unnecessary txn.ID != nil checks

Transactional requests are never sent with nil txn ids, so these checks
were unnecessary.

andreimatei committed:
roachpb: make HandledRetryableError.{TxnID,Transaction} non-nullable

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17299)
<!-- Reviewable:end -->

andreimatei commented:
This is based on @nvanbenschoten's #15024

cc @bdarnell for the first commit in particular

benesch commented:
Major 👩‍🔬 🐶, but: 🎉 

These semantics would have saved me so much pain.

nvanbenschoten commented:
Reviewed 1 of 1 files at r1, 33 of 33 files at r2, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/internal/client/txn.go, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqOZ6SEoRU1rgbt2Icp:-KqOZ6SEoRU1rgbt2Icq:b-4ie9by) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/internal/client/txn.go#L80)):*
> ```Go
> 		// response was meant for any incarnation of this transaction.
> 		//
> 		// TODO(andrei): This concept of the history of IDs seems problematic to me:
> ```

Note that this was not added to permit concurrent access to `Txn`, it was added to address the existing issue of identifying (possibly delayed) errors sent to previous incarnations of `Txn`.

---

*[pkg/internal/client/txn.go, line 104 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqOZdbacfB9oYjMdhyK:-KqOZdbacfB9oYjMdhyL:bpw2eu6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/internal/client/txn.go#L104)):*
> ```Go
> // NewTxn returns a new txn.
> func NewTxn(db *DB) *Txn {
> 	return NewTxnWithProto(db, *roachpb.NewTransaction(
> ```

Dereferencing a pointer immediately after allocating makes me sad. That said, I think there was pushback against creating a `roachpb.MakeTransaction` function.

---

*[pkg/internal/client/txn.go, line 715 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqOl82qsB1UW05Cy-K3:-KqOl82qsB1UW05Cy-K4:bcq1x4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/internal/client/txn.go#L715)):*
> ```Go
> // used.
> //
> // It is undefined to call Commit concurrently with any call to Exec. Since Exec
> ```

Looks like this block was duplicated.

---

*[pkg/roachpb/data.go, line 840 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqOoFnEIMROdZ__hmxt:-KqOoFnEIMROdZ__hmxu:b-goojgl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/roachpb/data.go#L840)):*
> ```Go
> // TxnIDEqual returns whether the transaction IDs are equal.
> // The IDs are passed by reference for efficiency, but they can't be nil.
> func TxnIDEqual(a, b *uuid.UUID) bool {
> ```

Why keep this when it can be replaced by `*a == *b`?

---

*[pkg/roachpb/data.go, line 1020 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqOofu8vxvkWKqP0s0M:-KqOofu8vxvkWKqP0s0N:b-x67e08) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/roachpb/data.go#L1020)):*
> ```Go
> 			txn.Name,
> 			nil,                // baseKey
> 			NormalUserPriority, // overwritten below
> ```

I'd add a comment addressing why we can't just use `errTxnPri` here.

---

*[pkg/roachpb/errors.proto, line 414 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqP24djqrUPH9DjGH1b:-KqP24djqrUPH9DjGH1c:b-lzrouk) ([raw file](https://github.com/cockroachdb/cockroach/blob/415f2182bf9dd010f5cf8bdde1828ee6da27a4c9/pkg/roachpb/errors.proto#L414)):*
> ```protobuf
>   // this against the ID of its transaction and make sure the retryable error
>   // is meant for its level and didn't escape from some inner transaction.
>   optional bytes txn_id = 2 [
> ```

Out of curiosity, what happens if we see an old proto with a null field on the wire after we make the field non-nullable?

---

*[pkg/sql/parser/builtins.go, line 1701 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqOt_qGChMiw_z6wB9F:-KqOt_qGChMiw_z6wB9G:bj7ugow) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/sql/parser/builtins.go#L1701)):*
> ```Go
> 	},
> 
> 	// force_retry returns a retryable error during the specified duration, and 0
> ```

I'd add a bit more to this. Something like: `If force_retry is called during the specified interval from the beginning of the transaction it returns a retryable error. If not, a 0 datum is returned instead of an error.`

---

*[pkg/storage/replica_command.go, line 117 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqO6ivzb4zpbpBLJoqa:-KqO6ivzb4zpbpBLJoqb:brynvpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/593c8077dfca88485843a4e86808edb8d255a26c/pkg/storage/replica_command.go#L117)):*
> ```Go
> 	}
> 	if header.Txn != nil {
> 		header.Txn.AssertInitialized(context.TODO())
> ```

Does this first commit compile? Looks like the `AssertInitialized` method was added in the second commit.

---

*[pkg/storage/engine/enginepb/mvcc.go, line 25 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqO6xXePucyvrseMmft:-KqO6xXePucyvrseMmfu:b-x2k2pi) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/storage/engine/enginepb/mvcc.go#L25)):*
> ```Go
> 		return id.Short()
> 	}
> 	// There shouldn't be any nil IDs any more.
> ```

Should we just change this method to `return id.Short()` and permit the panic on nil then?

---

*[pkg/storage/engine/enginepb/mvcc.proto, line 40 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqO8PoZPfxTF3KqT5U0:-KqO8PoZPfxTF3KqT5U1:b-6rc70g) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/storage/engine/enginepb/mvcc.proto#L40)):*
> ```protobuf
> 
>   // id is a unique UUID value which identifies the transaction.
>   // This field is always filled in.
> ```

As long as the `roachpb.NewTransaction` constructor is used. I'd point to that ctor here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 36 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/internal/client/txn.go, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqOZ6SEoRU1rgbt2Icp:-KqUkBpkGa13dZ_peRox:b-9vp97e) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/internal/client/txn.go#L80)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Note that this was not added to permit concurrent access to `Txn`, it was added to address the existing issue of identifying (possibly delayed) errors sent to previous incarnations of `Txn`.
</blockquote></details>

yeah I'm aware, but I think handling of a delayed retryable error should be no different from errors from parallel statements. Basically I think we should make it so that just the current id is necessary. Wouldn't you agree?

---

*[pkg/internal/client/txn.go, line 104 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqOZdbacfB9oYjMdhyK:-KqUlvBtrTsTgZEHs25H:bmuaplc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/internal/client/txn.go#L104)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Dereferencing a pointer immediately after allocating makes me sad. That said, I think there was pushback against creating a `roachpb.MakeTransaction` function.
</blockquote></details>

yeah... I don't know what to say.

---

*[pkg/internal/client/txn.go, line 715 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqOl82qsB1UW05Cy-K3:-KqUmkgohFXB7M6XC1dN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/internal/client/txn.go#L715)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Looks like this block was duplicated.
</blockquote></details>

Done.

---

*[pkg/roachpb/data.go, line 840 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqOoFnEIMROdZ__hmxt:-KqUmrf_NECBthhhRMhM:b-pxiig9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/roachpb/data.go#L840)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why keep this when it can be replaced by `*a == *b`?
</blockquote></details>

I'd keep it a little bit for the assertion

---

*[pkg/roachpb/data.go, line 1020 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqOofu8vxvkWKqP0s0M:-KqUnhYWwPM8XNy6Fd7H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/roachpb/data.go#L1020)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'd add a comment addressing why we can't just use `errTxnPri` here.
</blockquote></details>

Done.

---

*[pkg/roachpb/errors.proto, line 414 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqP24djqrUPH9DjGH1b:-KqUnoa6fl7q8J7KwV6g:b-k2n2d7) ([raw file](https://github.com/cockroachdb/cockroach/blob/415f2182bf9dd010f5cf8bdde1828ee6da27a4c9/pkg/roachpb/errors.proto#L414)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Out of curiosity, what happens if we see an old proto with a null field on the wire after we make the field non-nullable?
</blockquote></details>

these protos don't go on the wire; this is created by the `TxnCoordSender`.
If they did, I believe the server would get an id with all zeros.

---

*[pkg/sql/parser/builtins.go, line 1701 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqOt_qGChMiw_z6wB9F:-KqUolZl_4igUMcmSOBq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/sql/parser/builtins.go#L1701)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'd add a bit more to this. Something like: `If force_retry is called during the specified interval from the beginning of the transaction it returns a retryable error. If not, a 0 datum is returned instead of an error.`
</blockquote></details>

Done.

---

*[pkg/storage/replica_command.go, line 117 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqO6ivzb4zpbpBLJoqa:-KqUoy373_C3jC-anPSH:bfol50d) ([raw file](https://github.com/cockroachdb/cockroach/blob/593c8077dfca88485843a4e86808edb8d255a26c/pkg/storage/replica_command.go#L117)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this first commit compile? Looks like the `AssertInitialized` method was added in the second commit.
</blockquote></details>

nup, it was added in the first commit

---

*[pkg/storage/engine/enginepb/mvcc.go, line 25 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqO6xXePucyvrseMmft:-KqUp7QhLd0XQZDVBWX2:b-wxx0z5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/storage/engine/enginepb/mvcc.go#L25)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we just change this method to `return id.Short()` and permit the panic on nil then?
</blockquote></details>

ummm I'm a little concerned about Transactions coming over the wire with nils. 

---

*[pkg/storage/engine/enginepb/mvcc.proto, line 40 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqO8PoZPfxTF3KqT5U0:-KqV0lC5hYRxjhnCxuUL:buro181) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/storage/engine/enginepb/mvcc.proto#L40)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

As long as the `roachpb.NewTransaction` constructor is used. I'd point to that ctor here.
</blockquote></details>

well saying that anything is always filled in in go is conditional on some assumptions, and I don't think we generally document what those are.
I don't think this is the right place to point to some ctor.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 5 of 36 files at r4, 33 of 33 files at r5, 5 of 5 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/internal/client/txn.go, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqOZ6SEoRU1rgbt2Icp:-KqYqKF7BRnU49_4jQTZ:b114gdl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/internal/client/txn.go#L80)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

yeah I'm aware, but I think handling of a delayed retryable error should be no different from errors from parallel statements. Basically I think we should make it so that just the current id is necessary. Wouldn't you agree?
</blockquote></details>

In light of our discussion on Txn abortion and how we'd like to stop changing Txn IDs, I agree with you.

---

*[pkg/roachpb/errors.proto, line 414 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqP24djqrUPH9DjGH1b:-KqYrRLWO_4UTjicotzw:b3er01o) ([raw file](https://github.com/cockroachdb/cockroach/blob/415f2182bf9dd010f5cf8bdde1828ee6da27a4c9/pkg/roachpb/errors.proto#L414)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

these protos don't go on the wire; this is created by the `TxnCoordSender`.
If they did, I believe the server would get an id with all zeros.
</blockquote></details>

Ok, just verifying that this change won't need a migration path.

---

*[pkg/storage/replica_command.go, line 117 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqO6ivzb4zpbpBLJoqa:-KqYsmHiz_eeRr6ML8nI:bn0dlp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/593c8077dfca88485843a4e86808edb8d255a26c/pkg/storage/replica_command.go#L117)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nup, it was added in the first commit
</blockquote></details>

Are you sure? I'm not seeing that.

---

*[pkg/storage/engine/enginepb/mvcc.go, line 25 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqO6xXePucyvrseMmft:-KqYsys5-pBNZ-fEl3HG:b-d2jhbe) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/storage/engine/enginepb/mvcc.go#L25)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

ummm I'm a little concerned about Transactions coming over the wire with nils. 
</blockquote></details>

Don't we want it to panic then?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
@tschottdorf @bdarnell : pushed by @nvanbenschoten, I want to discuss again making `TxnMeta.ID` non-nillable (in a different PR). That ID has already been filled in for IDs in requests going through the `TxnCoordSender` - even reads before writes - that's old behavior. Thus, there's no risk of new nodes (non-nil) sending id `0000` to old nodes not prepared to handle that. So, I believe the only reason for keeping the ID be a pointer is to save memory when copying a `TxnMeta` (and a `Transaction` around). This is something Tobi has raised; I wouldn't have thought twice about it. But do we really copy `Transactions` all that much?
I think we should make this change. What do y'all think?

---

Review status: 1 of 37 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica_command.go, line 117 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqO6ivzb4zpbpBLJoqa:-KqZ62bCP5IHf1mcmE4N:b-6w1yw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/593c8077dfca88485843a4e86808edb8d255a26c/pkg/storage/replica_command.go#L117)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Are you sure? I'm not seeing that.
</blockquote></details>

you're right, sorry. Fixed.

---

*[pkg/storage/engine/enginepb/mvcc.go, line 25 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqO6xXePucyvrseMmft:-KqZ6atY1Jzl5_a5n4l8:b-qa9ayv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/storage/engine/enginepb/mvcc.go#L25)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Don't we want it to panic then?
</blockquote></details>

ok, made it just `id.Short()`. This was the wrong place to chicken out.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-:-KqZA2loQoNp-br-FFdj:btxq51o)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 37 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-:-KqZFI2o386Pg9qK6Hlu:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 36 files at r4, 31 of 34 files at r8, 5 of 5 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/internal/client/db_test.go, line 342 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqxjKnsigt_X0LvuoeB:-KqxjKntJsIMsZowaN4o:b-ymvf4c) ([raw file](https://github.com/cockroachdb/cockroach/blob/4975127cae3c5904529b2d5eff5129403ddf23e3/pkg/internal/client/db_test.go#L342)):*
> ```Go
> 
> 	if err := db.Txn(context.TODO(), func(ctx context.Context, txn *client.Txn) error {
> 		// Remove the txn ID, which will be added to the name if non-nil.
> ```

Since no real transaction will have a nil id, maybe we should set it to a hard-coded UUID value instead. 

---

*[pkg/roachpb/data.go, line 837 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kqxm-T1sue8fVfRkglW:-Kqxm-T1sue8fVfRkglX:b7092i2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4975127cae3c5904529b2d5eff5129403ddf23e3/pkg/roachpb/data.go#L837)):*
> ```Go
> 
> // TxnIDEqual returns whether the transaction IDs are equal.
> // The IDs are passed by reference for efficiency, but they can't be nil.
> ```

This seems like very little efficiency: a pointer is 8 bytes, and the full uuid is 16. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-:-Kqxnq7tds6EEQlgomBR:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
roachpb: get rid of TxnIDEqual()

Since IDs can't be nil any more, this function is replaced with directly
using ==.

andreimatei commented:
Ben, any thoughts on making `TxnMeta.ID` non-nillable? See question above ^.
Since I've gotten rid of `TxnIDEqual()`, I'd like to do that even more now because otherwise there's some risk that someone might compare pointers (instead of their values) when intending to compare IDs.

---

Review status: 0 of 40 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/internal/client/db_test.go, line 342 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqxjKnsigt_X0LvuoeB:-KqyqYAQppXUeGgxAcT-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4975127cae3c5904529b2d5eff5129403ddf23e3/pkg/internal/client/db_test.go#L342)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Since no real transaction will have a nil id, maybe we should set it to a hard-coded UUID value instead. 
</blockquote></details>

Done.

---

*[pkg/roachpb/data.go, line 837 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kqxm-T1sue8fVfRkglW:-KqyqleT-14WAKcZc5Lg:bskg83z) ([raw file](https://github.com/cockroachdb/cockroach/blob/4975127cae3c5904529b2d5eff5129403ddf23e3/pkg/roachpb/data.go#L837)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This seems like very little efficiency: a pointer is 8 bytes, and the full uuid is 16. 
</blockquote></details>

the real reason here was the nil panic. But since I'd also gotten ridden of other similar panics since writing this, I've added a commit that removes the TxnIDEqual() function in favor of using `==` directly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-:-KqyxM3o-qyZ1irubgbM:bej8jat)*
<!-- Sent from Reviewable.io -->

tamird commented:
+1 to making `TxnMeta.ID` non-nullable.

---

Reviewed 1 of 1 files at r1, 13 of 33 files at r2, 37 of 37 files at r10, 31 of 34 files at r11, 5 of 5 files at r12, 8 of 8 files at r13.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/storageccl/add_sstable_test.go, line 239 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqM1xT1wPwTMqPDgC9z:-KqM1xT1wPwTMqPDgCA-:b-6enw82) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/ccl/storageccl/add_sstable_test.go#L239)):*
> ```Go
> 			txn,
> 		); err != nil {
> 			if _, isWriteIntentErr := err.(*roachpb.WriteIntentError); !isWriteIntentErr {
> ```

how come this error changed?

---

*[pkg/internal/client/db_test.go, line 344 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqyzGBBH5qO6CXfhL4d:-KqyzGBBH5qO6CXfhL4e:bb16743) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b4353feae25283eb4667139413d82826fc5ebc4/pkg/internal/client/db_test.go#L344)):*
> ```Go
> 
> 	if err := db.Txn(context.TODO(), func(ctx context.Context, txn *client.Txn) error {
> 		// Remove the txn ID, which will be added to the name if non-nil.
> ```

what is this comment?

---

*[pkg/internal/client/txn.go, line 104 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqM2PE12UHLBkKcFs3I:-KqM2PE12UHLBkKcFs3J:b-rkmxd4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/internal/client/txn.go#L104)):*
> ```Go
> // NewTxn returns a new txn.
> func NewTxn(db *DB) *Txn {
> 	return NewTxnWithProto(db, *roachpb.NewTransaction(
> ```

if it's safe to deref it, why does `NewTransaction` return a pointer?

---

*[pkg/kv/txn_coord_sender.go, line 838 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqM4bnx1T6w3AR7E5-s:-KqM4bnx1T6w3AR7E5-t:byqqo76) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/kv/txn_coord_sender.go#L838)):*
> ```Go
> 	} else {
> 		if pErr.TransactionRestart != roachpb.TransactionRestart_NONE {
> 			errTxnID := *pErr.GetTxn().ID // The ID of the txn that needs to be restarted.
> ```

why the copy and extra alloc?

---

*[pkg/roachpb/data.go, line 748 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqyyNpnEsFPJ73iIduO:-KqyyNpnEsFPJ73iIduP:b-lb1gn3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c0d36a70c5f63c5e57b1265dcd033643fc07e78/pkg/roachpb/data.go#L748)):*
> ```Go
> // AssertInitialized crashes if the transaction is not initialized.
> func (t *Transaction) AssertInitialized(ctx context.Context) {
> 	if (t.ID == nil) || (t.OrigTimestamp == hlc.Timestamp{}) || (t.Timestamp == hlc.Timestamp{}) {
> ```

some of these parenthesis are unnecessary. try: `if t.ID == nil || t.OrigTimestamp == (hlc.Timestamp{}) || t.Timestamp == (hlc.Timestamp{}) {`

---

*[pkg/roachpb/errors.proto, line 415 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kqyz_VRIxKQ-NBOwZ4E:-Kqyz_VRIxKQ-NBOwZ4F:bbtccag) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfc06febaa47efc21383912163258b4f9f609209/pkg/roachpb/errors.proto#L415)):*
> ```protobuf
>     (gogoproto.customtype) = "github.com/cockroachdb/cockroach/pkg/util/uuid.UUID",
>     (gogoproto.customname) = "TxnID",
>     (gogoproto.nullable) = false];
> ```

can you s/mandatory/non-nullable/ the commit message?

---

*[pkg/storage/engine/enginepb/mvcc.proto, line 40 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqM1oYAZHJNnOOUnBN1:-KqM1oYBlL66jLO8p2j3:b-vf5ap7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/storage/engine/enginepb/mvcc.proto#L40)):*
> ```protobuf
> 
>   // id is a unique UUID value which identifies the transaction.
>   // This field is always filled in.
> ```

should it be nullable = false? (does that compile?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-:-Kqz-HMu456jJgOCjwa_:bx7y9zs)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/storageccl/add_sstable_test.go, line 239 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqM1xT1wPwTMqPDgC9z:-Kqz2TacYo4uVuxE-vrY:bqlkv2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/ccl/storageccl/add_sstable_test.go#L239)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

how come this error changed?
</blockquote></details>

the old error was a "replay protection" that was incorrectly triggered because the txn didn't have the id correctly filled in

---

*[pkg/internal/client/db_test.go, line 344 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqyzGBBH5qO6CXfhL4d:-Kqz30Xh7RG_r2kFkbKg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b4353feae25283eb4667139413d82826fc5ebc4/pkg/internal/client/db_test.go#L344)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what is this comment?
</blockquote></details>

Done.

---

*[pkg/internal/client/txn.go, line 104 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqM2PE12UHLBkKcFs3I:-Kqz8_NOS5Oxna9SuA7K:b-68svvx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/internal/client/txn.go#L104)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

if it's safe to deref it, why does `NewTransaction` return a pointer?
</blockquote></details>

well I don't think that's the test exactly, but I've switched to `MakeTransaction()`.
`Transactions` are used both by value and by pointer in places, but by value is more common.

---

*[pkg/kv/txn_coord_sender.go, line 838 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqM4bnx1T6w3AR7E5-s:-Kqz3iu9b-BltydjGaMT:b5atczs) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/kv/txn_coord_sender.go#L838)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why the copy and extra alloc?
</blockquote></details>

why do you think there's an alloc here?

---

*[pkg/roachpb/data.go, line 748 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqyyNpnEsFPJ73iIduO:-Kqz47uB2oUBma7fzEu2:b-hz698y) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c0d36a70c5f63c5e57b1265dcd033643fc07e78/pkg/roachpb/data.go#L748)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

some of these parenthesis are unnecessary. try: `if t.ID == nil || t.OrigTimestamp == (hlc.Timestamp{}) || t.Timestamp == (hlc.Timestamp{}) {`
</blockquote></details>

ah good, thanks

---

*[pkg/roachpb/errors.proto, line 415 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kqyz_VRIxKQ-NBOwZ4E:-KqzCcWSHOInpnaHRkLU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfc06febaa47efc21383912163258b4f9f609209/pkg/roachpb/errors.proto#L415)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can you s/mandatory/non-nullable/ the commit message?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 39 of 39 files at r14, 41 of 41 files at r15, 5 of 5 files at r16, 8 of 8 files at r17.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/kv/txn_coord_sender.go, line 838 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqM4bnx1T6w3AR7E5-s:-KqzGAA_PDsDLhJa7xyW:b-g6w4xm) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05057b8ad0265791dc052d7bdb8658efb45e389/pkg/kv/txn_coord_sender.go#L838)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

why do you think there's an alloc here?
</blockquote></details>

This comment was left on an old iteration which had this code:
```
			errTxnID := *pErr.GetTxn().ID // The ID of the txn that needs to be restarted.
			if !roachpb.TxnIDEqual(&errTxnID, &txnID) {
```

Looks fine now.

---

*[pkg/roachpb/data.go, line 988 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqzGVycY144lyr-QZvi:-KqzGVycY144lyr-QZvj:b-49c8r3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f097708435bd19b09937cd901457cfad427d2e9/pkg/roachpb/data.go#L988)):*
> ```Go
> 
> 	txnClone := txn.Clone()
> 	txn = &txnClone
> ```

why is this a pointer? we deref it at the end of this method!

---

*[pkg/roachpb/errors.proto, line 415 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kqyz_VRIxKQ-NBOwZ4E:-KqzFOXFQ3kNglrj371-:bs7ntvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfc06febaa47efc21383912163258b4f9f609209/pkg/roachpb/errors.proto#L415)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Eh, sorry - didn't notice that this comment also makes `Transaction` non-nullable. Could you include that in the message as well?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299)*
<!-- Sent from Reviewable.io -->

tamird commented on pkg/sql/executor.go:
```diff
@@ -759,9 +759,7 @@ func (e *Executor) execParsed(
 
 		inTxn := txnState.State() != NoTxn
-		execOpt := client.TxnExecOptions{
-			AssignTimestampImmediately: true,
```
did you forget to actually remove `AssignTimestampImmediately` from the struct, though?

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Making TxnMeta.ID non-nullable seems OK to me. 

---

Reviewed 1 of 37 files at r10, 34 of 41 files at r15, 4 of 5 files at r16, 8 of 8 files at r17.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-:-KqzVaadXhgqykTNjPcr:b-iqtdwx)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
roachpb, all: make TxnMeta.ID non-nullable

Now that ID's are always filled in by the Transaction ctor, the proto
field can be non-nullable. The field had always been filled in when
sending transactional requests - so no protos with a nil value should
have been saved anywhere.

The commit required changing the golden hash value in the
protosBelowRaft test because the auto-generated code producing random
filled protos for MVCCMetadata has changed (which includes a TxnMeta).
Note that the TxnMeta is nullable in the proto, and in the cases where
MVCCMetadata is written below Raft, it is not filled in (but the test
doesn't know this).

andreimatei commented:
@ tamird, I had removed `AssignTimestampImmediately` from the struct.

Added a commit for non-nullable ID.

---

Review status: 0 of 62 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/data.go, line 988 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-KqzGVycY144lyr-QZvi:-Kr1gYmJlPLgHGOJhyx_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f097708435bd19b09937cd901457cfad427d2e9/pkg/roachpb/data.go#L988)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why is this a pointer? we deref it at the end of this method!
</blockquote></details>

Done.

---

*[pkg/roachpb/errors.proto, line 415 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kqyz_VRIxKQ-NBOwZ4E:-Kr1hS-d-gs8HZfsHfgJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfc06febaa47efc21383912163258b4f9f609209/pkg/roachpb/errors.proto#L415)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Eh, sorry - didn't notice that this comment also makes `Transaction` non-nullable. Could you include that in the message as well?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-:-Kr1gnPFrEadiW2nA6fp:b-d3ea74)*
<!-- Sent from Reviewable.io -->

tamird commented:
:+1: but merge conflict :lock: 

---

Reviewed 46 of 46 files at r18, 41 of 41 files at r19, 5 of 5 files at r20, 8 of 8 files at r21, 32 of 32 files at r22.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/internal/client/txn_test.go, line 176 at r22](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kr8EohiqwpIhuNKtdtC:-Kr8EohiqwpIhuNKtdtD:buucgzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/002a50621c4ec1e2f1f0835106880fa8335d847d/pkg/internal/client/txn_test.go#L176)):*
> ```Go
> 		{hlc.Timestamp{WallTime: 10, Logical: 1}, hlc.Timestamp{WallTime: 20, Logical: 1}},
> 		{hlc.Timestamp{WallTime: 20, Logical: 1}, hlc.Timestamp{WallTime: 20, Logical: 1}},
> 		{hlc.Timestamp{WallTime: 20, Logical: 1}, hlc.Timestamp{WallTime: 19, Logical: 0}},
> ```

how come this changed? just didn't want to leave it at zero, or did the zero timestamp acquire new meaning in this commit?

---

*[pkg/roachpb/batch.go, line 60 at r22](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kr8FTjW7NFVgr7nonJL:-Kr8FTjW7NFVgr7nonJM:bo7tiu7) ([raw file](https://github.com/cockroachdb/cockroach/blob/002a50621c4ec1e2f1f0835106880fa8335d847d/pkg/roachpb/batch.go#L60)):*
> ```Go
> 		return
> 	}
> 	otherTxn.AssertInitialized(context.TODO())
> ```

kind of a random place, no?

---

*[pkg/roachpb/data.go, line 869 at r22](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kr8Fb31pP-KRp-ZnwLu:-Kr8Fb31pP-KRp-ZnwLv:b-nmsmq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/002a50621c4ec1e2f1f0835106880fa8335d847d/pkg/roachpb/data.go#L869)):*
> ```Go
> 		return
> 	}
> 	o.AssertInitialized(context.TODO())
> ```

this one, too.

---

*[pkg/storage/push_txn_queue.go, line 248 at r22](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kr8Fv-26BpuB2hamReF:-Kr8Fv-26BpuB2hamReG:bajeqql) ([raw file](https://github.com/cockroachdb/cockroach/blob/002a50621c4ec1e2f1f0835106880fa8335d847d/pkg/storage/push_txn_queue.go#L248)):*
> ```Go
> // pending waiters.
> func (ptq *pushTxnQueue) UpdateTxn(ctx context.Context, txn *roachpb.Transaction) {
> 	txn.AssertInitialized(context.TODO())
> ```

you have a context here, you don't need to use TODO.

---

*[pkg/storage/replica.go, line 4537 at r22](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kr8G6TcF6sdo6zDolDN:-Kr8G6TcF6sdo6zDolDO:bjkijw6) ([raw file](https://github.com/cockroachdb/cockroach/blob/002a50621c4ec1e2f1f0835106880fa8335d847d/pkg/storage/replica.go#L4537)):*
> ```Go
> 			// Restore the original txn's Writing bool if the error specifies a
> 			// transaction.
> 			txn := result.Local.Err.GetTxn()
> ```

```
if txn := result.Local.Err.GetTxn(); txn != nil {
  if ba.Txn == nil {
    log.Fatalf(ctx, "error had a txn but batch is non-transactional. Err txn: %s", txn)
  } else if txn.ID == ba.Txn.ID {
    txn.Writing = wasWriting
  }
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-:-Kr8GgKjUsgNQrlFgVWj:bufl2th)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Reviewed 26 of 41 files at r19, 4 of 5 files at r20, 31 of 32 files at r22.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/storage/below_raft_protos_test.go, line 68 at r22](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kr8oWPq3KlMD8B40wYz:-Kr8oWPq3KlMD8B40wZ-:b-748vvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/002a50621c4ec1e2f1f0835106880fa8335d847d/pkg/storage/below_raft_protos_test.go#L68)):*
> ```Go
> 		populatedConstructor: func(r *rand.Rand) proto.Message { return enginepb.NewPopulatedMVCCMetadata(r, false) },
> 		emptySum:             7551962144604783939,
> 		populatedSum:         8188088666885167358,
> ```

I'm surprised that `populatedSum` changed but `emptySum` didn't. 

Anyway, this is the part of this change that needs the most scrutiny. It's safe iff every *persisted* MVCCMetadata with a non-nil TxnMeta also has a non-zero ID (which should be true). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-:-Kr8paQTeN8XJGuOJ5jg:bb6ywmx)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 62 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/internal/client/txn_test.go, line 176 at r22](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kr8EohiqwpIhuNKtdtC:-KrCLJgLN6hRjoySE30H:bdc8b0y) ([raw file](https://github.com/cockroachdb/cockroach/blob/002a50621c4ec1e2f1f0835106880fa8335d847d/pkg/internal/client/txn_test.go#L176)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

how come this changed? just didn't want to leave it at zero, or did the zero timestamp acquire new meaning in this commit?
</blockquote></details>

with this commit, zero is not an acceptable timestamp any more - an assertion about the Transaction proto being uninitialized fires.

---

*[pkg/roachpb/batch.go, line 60 at r22](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kr8FTjW7NFVgr7nonJL:-KrCLuFQRVB6EnhRiRVD:byt99ul) ([raw file](https://github.com/cockroachdb/cockroach/blob/002a50621c4ec1e2f1f0835106880fa8335d847d/pkg/roachpb/batch.go#L60)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

kind of a random place, no?
</blockquote></details>

it's not completely random - this one and `transaction.Update()` are places where one txn pollinates another so I figured an assertion here is reasonable

---

*[pkg/storage/below_raft_protos_test.go, line 68 at r22](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kr8oWPq3KlMD8B40wYz:-KrCPPWKO5Nhae8yCqVs:b6rgahi) ([raw file](https://github.com/cockroachdb/cockroach/blob/002a50621c4ec1e2f1f0835106880fa8335d847d/pkg/storage/below_raft_protos_test.go#L68)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'm surprised that `populatedSum` changed but `emptySum` didn't. 

Anyway, this is the part of this change that needs the most scrutiny. It's safe iff every *persisted* MVCCMetadata with a non-nil TxnMeta also has a non-zero ID (which should be true). 
</blockquote></details>

`emptySum` didn't change because `MVCCMetadata` has a nullable `TxnMeta` (like you've seen) - so the tree is cut above our change to `TxnMeta.ID`.

---

*[pkg/storage/push_txn_queue.go, line 248 at r22](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kr8Fv-26BpuB2hamReF:-KrCQfFjQaSxXHCg8Wgn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/002a50621c4ec1e2f1f0835106880fa8335d847d/pkg/storage/push_txn_queue.go#L248)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

you have a context here, you don't need to use TODO.
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 4537 at r22](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kr8G6TcF6sdo6zDolDN:-KrCRKQkHJeY-du0L3aQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/002a50621c4ec1e2f1f0835106880fa8335d847d/pkg/storage/replica.go#L4537)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

```
if txn := result.Local.Err.GetTxn(); txn != nil {
  if ba.Txn == nil {
    log.Fatalf(ctx, "error had a txn but batch is non-transactional. Err txn: %s", txn)
  } else if txn.ID == ba.Txn.ID {
    txn.Writing = wasWriting
  }
}
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 25 of 41 files at r24, 5 of 6 files at r25, 32 of 32 files at r27.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/roachpb/batch.go, line 60 at r22](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kr8FTjW7NFVgr7nonJL:-KrCj6BnZWdQ2OYlHV30:b-rdkrmo) ([raw file](https://github.com/cockroachdb/cockroach/blob/002a50621c4ec1e2f1f0835106880fa8335d847d/pkg/roachpb/batch.go#L60)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

it's not completely random - this one and `transaction.Update()` are places where one txn pollinates another so I figured an assertion here is reasonable
</blockquote></details>

Consider adding comments to that effect, because it looks very random.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299)*
<!-- Sent from Reviewable.io -->

knz commented:
typo in title of first commit.

andreimatei commented:
Fixed commit title.

---

Review status: 0 of 62 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/roachpb/batch.go, line 60 at r22](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kr8FTjW7NFVgr7nonJL:-KrH_p3HP3AFom1ckjIW:bhs0tf3) ([raw file](https://github.com/cockroachdb/cockroach/blob/002a50621c4ec1e2f1f0835106880fa8335d847d/pkg/roachpb/batch.go#L60)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Consider adding comments to that effect, because it looks very random.
</blockquote></details>

Well I'm considering it, but I don't exactly know what to write. A ton of places could have these assertions, but they don't because we don't put such assertions everywhere.
The places where I did put it are somewhat arbitrary (you might even call them random) - they have some properties like the fact that here one transaction influences another (so perhaps consequences of a non-initialized one would be less localized and thus harder to debug), and also transactions going through these `Update()` methods come from diverse places (returned from RPCs, etc). So one of the considerations is that this assertion might catch various problems. But these arguments are not convincing enough to put them in the code in my opinion.
Anyway, in short, these places are indeed pretty random, but random assertions seem fine to me. We could make them less random by sprinkling more of them (but there's diminishing returns; the most important thing is that there's a first assertion somewhere, anywhere) or by moving them in more principled locations. Here I'll take suggestions, but otherwise I don't know good ones.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-:-KrH_lgumhbQSwb1kAzS:b595d66)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
🎉 🎉 🎉

tamird commented:
> they have some properties like the fact that here one transaction
influences another (so perhaps consequences of a non-initialized one would
be less localized and thus harder to debug), and also transactions going
through these Update() methods come from diverse places (returned from
RPCs, etc)

I think these considerations are useful enough to put in code, for no other
reason than giving future readers insight into your intent.

On Aug 11, 2017 14:54, "Andrei Matei" <notifications@github.com> wrote:

> Fixed commit title.
> ------------------------------
>
> Review status: 0 of 62 files reviewed at latest revision, 3 unresolved
> discussions, some commit checks pending.
> ------------------------------
>
> *pkg/roachpb/batch.go, line 60 at r22
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-Kr8FTjW7NFVgr7nonJL:-KrH_p3HP3AFom1ckjIW:bhs0tf3>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/002a50621c4ec1e2f1f0835106880fa8335d847d/pkg/roachpb/batch.go#L60>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> Consider adding comments to that effect, because it looks very random.
>
> Well I'm considering it, but I don't exactly know what to write. A ton of
> places could have these assertions, but they don't because we don't put
> such assertions everywhere.
> The places where I did put it are somewhat arbitrary (you might even call
> them random) - they have some properties like the fact that here one
> transaction influences another (so perhaps consequences of a
> non-initialized one would be less localized and thus harder to debug), and
> also transactions going through these Update() methods come from diverse
> places (returned from RPCs, etc). So one of the considerations is that this
> assertion might catch various problems. But these arguments are not
> convincing enough to put them in the code in my opinion.
> Anyway, in short, these places are indeed pretty random, but random
> assertions seem fine to me. We could make them less random by sprinkling
> more of them (but there's diminishing returns; the most important thing is
> that there's a first assertion somewhere, anywhere) or by moving them in
> more principled locations. Here I'll take suggestions, but otherwise I
> don't know good ones.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17299#-:-KrH_lgumhbQSwb1kAzS:b595d66>*
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/17299#issuecomment-321892132>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPPrJ1Nzy7Wqky_p4nbbuVgkZz2Rxks5sXKN2gaJpZM4OnIsR>
> .
>
